File: src/game/defense/tutorial/buildDefenseTutorialStaffDisplay.ts
import type { ChordVoicingStaffGroup } from '@/components/earTraining/ChordVoicingStaff';
import type { DefensePhraseJudgeState } from '@/game/defense/defensePhraseJudge';

export interface DefenseTutorialStaffDisplay {
  readonly groups: readonly ChordVoicingStaffGroup[];
  readonly activeGroupId: string | null;
  readonly correctPitchClassesByGroupId: ReadonlyMap<string, readonly number[]>;
}

export const buildDefenseTutorialStaffDisplay = (
  baseGroups: readonly ChordVoicingStaffGroup[],
  judge: DefensePhraseJudgeState,
  concertPitchClasses: readonly number[],
): DefenseTutorialStaffDisplay => {
  const correctPitchClassesByGroupId = new Map<string, readonly number[]>();
  let activeGroupId: string | null = null;

  const groups = baseGroups.map((group, index) => {
    if (group.isRest) {
      return group;
    }
    const stepIndex = index;
    const isComplete = judge.correctNoteIndices.has(stepIndex)
      || stepIndex < judge.targetStepIndex;
    const isActive = stepIndex === judge.targetStepIndex;
    if (isActive) {
      activeGroupId = group.id;
    }
    const pitchClass = concertPitchClasses[stepIndex];
    if (isComplete && pitchClass !== undefined) {
      correctPitchClassesByGroupId.set(group.id, [pitchClass]);
    }
    return {
      ...group,
      isActive,
      correctPitchClasses: isComplete && pitchClass !== undefined ? [pitchClass] : [],
    };
  });

  return {
    groups,
    activeGroupId,
    correctPitchClassesByGroupId,
  };
};

File: ios/Jazzify/Defense/Tutorial/BuildDefenseTutorialStaffDisplay.swift
import Foundation

struct DefenseTutorialStaffDisplay: Equatable, Sendable {
    let groups: [DefenseTutorialStaffGroup]
    let activeGroupId: UUID?
    let correctPitchClassesByGroupId: [UUID: Set<Int>]
}

enum BuildDefenseTutorialStaffDisplay {
    static func build(
        baseGroups: [DefenseTutorialStaffGroup],
        judge: DefensePhraseJudgeState,
        concertPitchClasses: [Int]
    ) -> DefenseTutorialStaffDisplay {
        var correctPitchClassesByGroupId: [UUID: Set<Int>] = [:]
        var activeGroupId: UUID?

        let groups = baseGroups.enumerated().map { index, group -> DefenseTutorialStaffGroup in
            if group.isRest {
                return group
            }
            let stepIndex = index
            let isComplete = judge.correctNoteIndices.contains(stepIndex)
                || stepIndex < judge.targetStepIndex
            let isActive = stepIndex == judge.targetStepIndex
            if isActive {
                activeGroupId = group.id
            }
            if isComplete, let pitchClass = concertPitchClasses[safe: stepIndex] {
                correctPitchClassesByGroupId[group.id] = [pitchClass]
            }
            return group
        }

        return DefenseTutorialStaffDisplay(
            groups: groups,
            activeGroupId: activeGroupId,
            correctPitchClassesByGroupId: correctPitchClassesByGroupId
        )
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        guard indices.contains(index) else { return nil }
        return self[index]
    }
}
